// svelte-use-form/models/form.js
import { FormControl } from "./formControl";
export class Form {
    constructor(initialData, notifyListeners) {
        for (const [k, v] of Object.entries(initialData !== null && initialData !== void 0 ? initialData : {})) {
            this._addFormControl(k, v.initial, v.validators, [], v.errorMap);
        }
        this._notifyListeners = notifyListeners;
    }
    // @ts-expect-error - Due to index signature
    get valid() {
        let valid = true;
        this.forEachFormControl((formControl) => {
            if (!formControl.valid)
                valid = false;
        });
        return valid;
    }
    // @ts-expect-error - Due to index signature
    get touched() {
        let touched = true;
        this.forEachFormControl((formControl) => {
            if (!formControl.touched)
                touched = false;
        });
        return touched;
    }
    // @ts-expect-error - Due to index signature
    get values() {
        let values = {};
        this.forEachFormControl((formControl, key) => {
            values[key] = formControl.value;
        });
        return values;
    }
    // @ts-expect-error - Due to index signature
    set touched(value) {
        this.forEachFormControl((formControl) => {
            formControl.touched = value;
        });
        this._notifyListeners();
    }
    /** Reset the whole form */
    // @ts-expect-error - Due to index signature
    reset() {
        this.forEachFormControl((formControl) => formControl.reset());
    }
    // @ts-expect-error - Due to index signature
    _addFormControl(name, initial, validators, elements, errorMap) {
        this[name] = new FormControl({
            value: initial !== null && initial !== void 0 ? initial : "",
            validators: validators !== null && validators !== void 0 ? validators : [],
            errorMap: errorMap !== null && errorMap !== void 0 ? errorMap : {},
            elements: elements !== null && elements !== void 0 ? elements : [],
            formRef: () => this,
        });
    }
    // @ts-expect-error - Due to index signature
    forEachFormControl(callback) {
        for (const [key, value] of Object.entries(this)) {
            if (value instanceof FormControl) {
                callback(value, key);
            }
        }
    }
}
